elixir/apps/api/lib/api/gateway/channel.ex

```diff
@@ -107,6 +259,72 @@ defmodule API.Gateway.Channel do
     end
   end
 
+  def handle_info(
+        {:authorize_flow, {channel_pid, socket_ref}, payload,
+         {opentelemetry_ctx, opentelemetry_span_ctx}},
+        socket
+      ) do
+    OpenTelemetry.Ctx.attach(opentelemetry_ctx)
+    OpenTelemetry.Tracer.set_current_span(opentelemetry_span_ctx)
+
+    %{
+      client_id: client_id,
+      resource_id: resource_id,
+      flow_id: flow_id,
+      authorization_expires_at: authorization_expires_at,
+      ice_credentials: ice_credentials,
+      preshared_key: preshared_key
+    } = payload
+
+    OpenTelemetry.Tracer.with_span "gateway.authorize_flow" do
+      :ok = Flows.subscribe_to_flow_expiration_events(flow_id)
+
+      Logger.debug("Gateway authorizes a new network flow",
+        flow_id: flow_id,
+        client_id: client_id,
+        resource_id: resource_id
+      )
+
+      client = Clients.fetch_client_by_id!(client_id, preload: [:actor])
+      resource = Resources.fetch_resource_by_id!(resource_id)
+
+      :ok = Resources.unsubscribe_from_events_for_resource(resource_id)
+      :ok = Resources.subscribe_to_events_for_resource(resource_id)
+
+      opentelemetry_headers = :otel_propagator_text_map.inject([])
+
+      ref =
+        encode_ref(socket, {
+          channel_pid,
+          socket_ref,
+          resource_id,
+          preshared_key,
+          ice_credentials,
+          opentelemetry_headers
+        })
+
+      push(socket, "authorize_flow", %{
+        ref: ref,
+        flow_id: flow_id,
+        actor: Views.Actor.render(client.actor),
+        resource: Views.Resource.render(resource),
+        gateway_ice_credentials: ice_credentials.gateway,
+        client: Views.Client.render(client, preshared_key),
```

Comment:
Would prefer if this was flattened out.

Reply:
Can we not flatten everything? If we do it will lead to more mess in tests and constructing that will either be a bunch of map merges or we won't have view modules for each entity individually.

Reply:
Oke, I'll adapt my deserialisation code :)